module/zfs/arc.c

```diff
@@ -5638,7 +5640,11 @@ arc_buf_access(arc_buf_t *buf)
 	}
 
 	kmutex_t *hash_lock = HDR_LOCK(hdr);
-	mutex_enter(hash_lock);
+	if (!mutex_tryenter(hash_lock)) {
+		mutex_exit(&buf->b_evict_lock);
+		ARCSTAT_BUMP(arcstat_access_skip);
+		return;
+	}
```

Comment:
I think think this needs a comment describing exactly the kind of deadlock we're trying to avoid here.